load X cursor images from PNG files

teleshoes opened the pull request:
previously, cursor images were stored directly in source with string initializers. the PNG files introduced here are pixel-identical to the string initializers, with the same color profile/depth (rgba32).

moving the images to external data files enables easier maintenance, possible user customization, and smaller binaries, as well as significantly smaller source (the PNG files are 20% the size of the string initializers).

performance seems to be identical. startup/initialization times are unchanged.

on a side note, the fingers' skin colors are not representative of all users in the mer community. it will now be easier for anyone to modify the source images to something neutral, or to allow customization on the command line (which is not in this pr).

teleshoes commented:
this is the pointer i have been using. it doesn't cover as much of the screen and allows for more accuracy, AND its not a big weird picture of a finger.

![blue_pointer](https://github.com/mer-qa/lipstick2vnc/assets/592062/4017ad9c-d624-49d5-b7cb-f8e2ee7b69f7)


![%usr%share%lipstick2vnc%cursor_pointer](https://github.com/mer-qa/lipstick2vnc/assets/592062/1287115f-c7d5-477f-8178-123d99aafde8)
![%usr%share%lipstick2vnc%cursor_pointer_touch](https://github.com/mer-qa/lipstick2vnc/assets/592062/7aab04e7-bd41-43ef-a689-43387867d0dc)

teleshoes commented:
not particularly, using the most widely used lib is obviously more sane. here are the reasons i didn't:

picopng is a lightweight single function that's easy to use, fast, and freely available with a permissive FOSS license, and used in a number of existing projects (so it's fairly standard, though of course nowhere near as standard as libpng.so). having no external dependencies was nice, especially when testing lossless-ness converting between the bitmasks and the png and vice-versa on my laptop and on the phone.

libpng is hard to use, heavyweight, the documentation reads like the novelization of one man's journey to learn every C++ anti-pattern, and my code had a small memory leak when i tried to get it to work on sailfish.

maybe this could be merged and a future pr could make it a bit more standards-friendly? i can try to get libpng working when i get some time (not sure when that will be).

Thaodan commented:
> picopng is a lightweight single function that's easy to use, fast, and freely available with a permissive FOSS license, and used in a number of existing projects (so it's fairly standard, though of course nowhere near as standard as libpng.so). having no external dependencies was nice,

I don't think we want in-tree libraries. It is easier to have external dependencies that can kept up to date separately. This resurgence on in-tree libraries is a scourge. 

Another option to use Qt's builtin image API's?

teleshoes commented:
hah, didnt think of using QImage. took about 5 min to write up, seems to work fine. ill push it after i run some more tests. thanks @Thaodan!

teleshoes commented:
ok, removed picopng from the commits, and added the new code as a separate new commit. i tested with the default PNGs included in this pr (generated from the masks), and with my custom ones. performance+memory usage seem unchanged.

teleshoes commented:
(@mlehtima fixed)

teleshoes commented:
@Thaodan any chance of getting this merged? i am happy to make any additional changes

pvuorela commented on src/screentovnc.cpp:
```diff
@@ -515,20 +515,82 @@ void ScreenToVnc::rfbProcessTrigger()
+            int px = y*width*bpp + x*bpp;
+            //unsigned char r = png[px + 0];
+            //unsigned char g = png[px + 1];
+            //unsigned char b = png[px + 2];
```
The commented out lines could be skipped too. The order is seen above.

pvuorela commented on utils/gen_cursor_pngs.c:
```diff
@@ -0,0 +1,42 @@
+#include <stdio.h>
```
The file here not used and suppose it makes more sense if wanting to recreate the png files from the old headers. At the final stage the headers are removed so maybe could remove the file on the same commit as the headers.

pvuorela commented on src/screentovnc.cpp:
```diff
@@ -515,20 +515,82 @@ void ScreenToVnc::rfbProcessTrigger()
+    info.width = width;
+    info.height = height;
+    info.pixel_data = png;
+    info.bitmask = (char*) malloc(sizeof(char) * width * height + 1);
```
Never freed. Maybe having a destructor on cursor_info could be a simple way? I.e. cursor_info having ownership here.

teleshoes replied:
simply putting the free in a ~cursor_info() breaks it, it gets free'd too early and the png is drawn with a random bitmask.
since the malloc in init_fingerPointers() is called from the ScreenToVnc() ctor, i just put the free in that already-present destructor.

pvuorela commented on src/screentovnc.h:
```diff
@@ -79,6 +84,9 @@ enum Orientation {
 static rfbCursor *pointerFingerTouchPtr;
+static cursor_info emptyMouseCursorInfo;
+static cursor_info pointerCursorInfo;
+static cursor_info pointerTouchCursorInfo;
```
Wondering is there any need for keeping these after rfbMakeXCursor() calls unless the cursor keeps some references on them. Otherwise these could be perhaps just temporary variables in init_fingerPointers()?

pvuorela reviewed (commented):
Could perhaps squash the commits a bit. E.g .now there's one commit for doing framework for loading pngs, but having "// TODO: implement reading PNG img data", and then later commit adding implementation with QImage. Could be merged.

On another ones there's first a commit creating png files and then another for installing them. These could be merged not only together but also to  the png reading commit.

teleshoes commented:
@pvuorela please let me know if these fixes are not sufficient. thanks!

pvuorela commented on rpm/lipstick2vnc.spec:
```diff
@@ -66,6 +66,9 @@ fi
 %files
 %defattr(-,root,root,-)
 %attr(2755, root, privileged) %{_bindir}/%{name}
+/usr/share/lipstick2vnc/cursor_empty.png
```
Should be %{_datadir} instead of /usr/share but I can adjust that. Could have a bunch of other cleanups as well.

pvuorela reviewed (approved):
I think this is good enough for merging.
